### PR TITLE
Add onlyIf(Provider<Boolean>) API

### PR DESCRIPTION
**Pull request by 6hundreds**
Part of https://github.com/gradle/gradle/issues/16080
Discussion https://gradle.slack.com/archives/C0WTFHWT1/p1680033872285069

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/Task.java [hunk 1]
```diff
@@ -379,6 +397,22 @@ public String determineName(Task c) {
...
+     * will be added to this task accordingly.</p>
+     *
+     * <p>The given predicate replaces all such predicates for this task.</p>
+     *
+     * @param onlyIfProvider provider to execute to determine if task should be run
+     *
+     * @since 8.2
+     */
+    @Incubating
+    void setOnlyIf(Provider<Boolean> onlyIfProvider);
```
Comment (bamboo) [inline subprojects/core-api/src/main/java/org/gradle/api/Task.java:414]: ❌ I wouldn't add `setOnlyIf`.

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/TaskOnlyIfIntegrationTest.groovy [hunk 2]
```diff
@@ -20,7 +20,33 @@ import org.gradle.integtests.fixtures.AbstractIntegrationSpec
...
+            tasks.register("dependent") {
+                it.${taskSkippingApi}(dependencyTask.map { false })
+            }
+        """
+
+        when:
+        run "dependent"
+
+        then:
+        result.assertTaskSkipped(":dependent")
```
Comment (bamboo) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/TaskOnlyIfIntegrationTest.groovy:43]: 👍

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java [hunk 3]
```diff
@@ -1069,6 +1094,17 @@ public List<ResourceLock> getSharedResources() {
         return getBuildServiceRegistry().getSharedResources(taskRequiredServices.getElements());
     }
 
+    private void addProviderTaskDependency(Provider<?> provider) {
+        // See DefaultTaskDependency.visitDependencies
+        if (provider instanceof ProviderInternal) {
+            ProviderInternal<?> providerInternal = (ProviderInternal<?>) provider;
+            ValueSupplier.ValueProducer producer = providerInternal.getProducer();
+            if (producer.isKnown()) {
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java:1102]: ❌ This is too early for this check. It should work to unconditionally add the provider to the `dependencies` collection.

File: subprojects/core/src/testFixtures/groovy/org/gradle/api/tasks/AbstractTaskTest.groovy [hunk 4]
```diff
@@ -168,6 +168,21 @@ abstract class AbstractTaskTest extends AbstractProjectBuilderSpec {
...
+        task.getOnlyIf().isSatisfiedBy(task)
+
+        when:
+        task.onlyIf(task.project.provider({ false }))
+
+        then:
+        !task.getOnlyIf().isSatisfiedBy(task)
+    }
+
+
```
Comment (bamboo) [inline subprojects/core/src/testFixtures/groovy/org/gradle/api/tasks/AbstractTaskTest.groovy:185]: 🎨 Remove spurious empty line.

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java [hunk 5]
```diff
@@ -321,6 +324,17 @@ public void run() {
         });
     }
 
+    @Override
+    public void onlyIf(final Provider<Boolean> provider) {
+        taskMutator.mutate("Task.onlyIf(Provider)", new Runnable() {
+            @Override
+            public void run() {
+                addProviderTaskDependency(provider);
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java:332]: ❓ It seems the task dependencies of this provider would be retained even if `setOnlyIf(() -> true)` is called afterwards, right?

**Review by bamboo — CHANGES_REQUESTED** [review]

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java (hunk 5 above)
Comment (6hundreds) [reply subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java:332]: Hm, yes, it is. Seems like we should remove dependency if `onlyIfSpec` doesn't contains `ProviderBackedSpec` with `TaskProvider` within?  Looking crappy

File: subprojects/core-api/src/main/java/org/gradle/api/Task.java [hunk 6]
```diff
@@ -344,6 +344,24 @@ public String determineName(Task c) {
...
+     * <p>You may add multiple such predicates. The task is skipped if any of the predicates return false.</p>
+     *
+     * <p>Typical usage:<code>myTask.onlyIf(project.provider({ isProductionEnvironment() }))</code></p>
+     *
+     * @param onlyIfProvider provider to execute to determine if task should be run
+     *
+     * @since 8.2
+     */
+    @Incubating
+    void onlyIf(Provider<Boolean> onlyIfProvider);
```
Comment (eskatos) [inline subprojects/core-api/src/main/java/org/gradle/api/Task.java:363]: :question: Why not add the overload taking an `onlyIfReason` string?

Comment (6hundreds) [reply subprojects/core-api/src/main/java/org/gradle/api/Task.java:363]: @eskatos Yeah, I'm going to add it also

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java (hunk 3 above)
Comment (bamboo) [reply subprojects/core/src/main/java/org/gradle/api/internal/AbstractTask.java:1102]: See https://github.com/gradle/gradle/pull/24612

**Comment by 6hundreds** [comment]
@bot-gradle test acc

**Comment by 6hundreds** [comment]
@bot-gradle test acc

**Comment by 6hundreds** [comment]
@bot-gradle test acc

**Comment by 6hundreds** [comment]
@bot-gradle test acc

**Pull request closed without merging**